File: src/main/java/io/github/arieslab/business/daos/utils/BusinessGeneric.java
package io.github.arieslab.business.daos.utils;

import org.hibernate.criterion.Order;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.transaction.annotation.Transactional;

import java.io.Serializable;
import java.util.List;

@Transactional
public class BusinessGeneric<T extends DAOGeneric, Y extends Serializable> {

    @Autowired
    protected T dao;

    /**
     * Returns the number of entities.
     *
     * @return the entity count
     */
    public int size() {
        return dao.size();
    }

    /**
     * Saves an entity.
     *
     * @param u the entity to save
     */
    public void save(Y u) {
        dao.save(u);
    }

    /**
     * Deletes an entity.
     *
     * @param u the entity to delete
     */
    public void delete(Y u) {
        dao.delete(u);
    }

    /**
     * Deletes an entity by its ID.
     *
     * @param id the entity ID
     */
    public void delete(Long id) {
        dao.delete(id);
    }

    /**
     * Finds entities with pagination, ordered by ID descending.
     *
     * @param first the offset
     * @param count the page size
     * @return list of entities
     */
    public List<Y> find(Long first, Long count) {
        return dao.findByHQL(Order.desc("id"), first.intValue(), count.intValue());
    }

    /**
     * Finds an entity by its ID.
     *
     * @param id the entity ID
     * @return the found entity, or null
     */
    public Y find(Long id) {
        return (Y) dao.findById(id);
    }

    /**
     * Lists all entities.
     *
     * @return list of all entities
     */
    public List<Y> listAll() {
        return dao.listAll();
    }

}

File: src/main/java/io/github/arieslab/business/daos/utils/BusinessGeneric.kt
package io.github.arieslab.business.daos.utils

import org.springframework.transaction.annotation.Transactional
import java.io.Serializable

@Transactional
open class BusinessGeneric<T : DAOGeneric<*>, Y : Serializable>(protected open var dao: T) {

    open fun size() = dao.size()

    open fun save(u: Y) {
        @Suppress("unchecked_cast")
        (dao as DAOGeneric<Y>).save(u)
    }

    open fun delete(u: Y) {
        @Suppress("unchecked_cast")
        (dao as DAOGeneric<Y>).delete(u)
    }

    open fun delete(id: Long) {
        @Suppress("unchecked_cast")
        (dao as DAOGeneric<Y>).delete(id)
    }

    @Suppress("unchecked_cast")
    open fun find(first: Long, count: Long): List<Y> {
        return (dao as DAOGeneric<Y>).findByHQL("id", false, first.toInt(), count.toInt())
    }

    @Suppress("unchecked_cast")
    open fun find(id: Long): Y? {
        return (dao as DAOGeneric<Y>).findById(id)
    }

    @Suppress("unchecked_cast")
    open fun listAll(): List<Y> {
        return (dao as DAOGeneric<Y>).listAll()
    }
}
